<a href="https://colab.research.google.com/github/Shreyasi-jcx3419/weightsandbiases/blob/main/Practicum_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Week 4 - Models and Experimentation

## Step 1 Training a model

For the purposes of this demo, we will be using this [adapted demo](https://www.datacamp.com/tutorial/xgboost-in-python) and training an XGBoost model, and then doing some experimentation and hyperparameter tuning.


If running this notebook locally, use the following steps to create virtual environment:
- Don't use past python 3.10
- To create virtual environment use "venv"

`python -m venv NAME`

- Try to avoid anaconda, poetry or similar package management platforms
- To install a package use pip

`python -m pip install <package-name>`

- once you are done working with this virtual environment, deactivate it with `deactivate`

### Install packages

In [1]:
!pip install wandb -qU

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 10.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.3/207.3 kB 9.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 267.1/267.1 kB 11.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.7/62.7 kB 5.2 MB/s eta 0:00:00


In [2]:
import xgboost as xgb
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error


### Import data

We will be using Diamonds dataset imported from Seaborn. It is also available on [Kaggle](https://www.kaggle.com/datasets/shivam2503/diamonds).

Read about the features by following the link. We will be predicting the price of diamonds.

In [3]:
diamonds = sns.load_dataset('diamonds')
diamonds.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

In [4]:
diamonds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   carat    53940 non-null  float64 
 1   cut      53940 non-null  category
 2   color    53940 non-null  category
 3   clarity  53940 non-null  category
 4   depth    53940 non-null  float64 
 5   table    53940 non-null  float64 
 6   price    53940 non-null  int64   
 7   x        53940 non-null  float64 
 8   y        53940 non-null  float64 
 9   z        53940 non-null  float64 
dtypes: category(3), float64(6), int64(1)
memory usage: 3.0 MB


In [5]:
diamonds.shape

(53940, 10)

In [6]:
X,y = diamonds.drop('price', axis=1), diamonds[['price']]

# For the cut, color and clarity use pandas category to enable XGBoost ability to deal with categorical data.

X['cut'] = X['cut'].astype('category')
X['color'] = X['color'].astype('category')
X['clarity'] = X['clarity'].astype('category')

### Split the data and train a model

In [7]:
# Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create DMatrix
dtrain = xgb.DMatrix(X_train, label=y_train, enable_categorical=True)
dtest = xgb.DMatrix(X_test, label=y_test, enable_categorical=True)

In [8]:
# Define hyperparameters
params = {"objective": "reg:squarederror", "tree_method": "gpu_hist"}

n = 100
model = xgb.train(
   params=params,
   dtrain=dtrain,
   num_boost_round=n,
)

/usr/local/lib/python3.10/dist-packages/xgboost/core.py:160: UserWarning: [03:57:43] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.

    E.g. tree_method = "hist", device = "cuda"

  warnings.warn(smsg, UserWarning)


In [9]:
# Define evaluation metrics - Root Mean Squared Error

predictions = model.predict(dtest)
rmse = mean_squared_error(y_test, predictions, squared=False)
print(f"RMSE: {rmse}")

RMSE: 532.8838153117543


/usr/local/lib/python3.10/dist-packages/xgboost/core.py:160: UserWarning: [03:57:44] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.

    E.g. tree_method = "hist", device = "cuda"

  warnings.warn(smsg, UserWarning)


### Incorporate validation

In [10]:
params = {"objective": "reg:squarederror", "tree_method": "gpu_hist"}
n = 100

# Create the validation set
evals = [(dtrain, "train"), (dtest, "validation")]

In [11]:
evals = [(dtrain, "train"), (dtest, "validation")]

model = xgb.train(
   params=params,
   dtrain=dtrain,
   num_boost_round=n,
   evals=evals,
   verbose_eval=10,
)

[0]	train-rmse:2859.49097	validation-rmse:2851.62630
[10]	train-rmse:550.99470	validation-rmse:571.16640


/usr/local/lib/python3.10/dist-packages/xgboost/core.py:160: UserWarning: [03:57:44] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.

    E.g. tree_method = "hist", device = "cuda"

  warnings.warn(smsg, UserWarning)


[20]	train-rmse:491.51435	validation-rmse:544.08058
[30]	train-rmse:464.38845	validation-rmse:537.01895
[40]	train-rmse:445.99106	validation-rmse:533.85127
[50]	train-rmse:430.36010	validation-rmse:532.90320
[60]	train-rmse:418.87898	validation-rmse:533.04629
[70]	train-rmse:409.66247	validation-rmse:533.58046
[80]	train-rmse:397.34048	validation-rmse:534.31963
[90]	train-rmse:389.94294	validation-rmse:532.61946
[99]	train-rmse:377.70831	validation-rmse:532.88383


In [12]:
# Incorporate early stopping
n = 10000


model = xgb.train(
   params=params,
   dtrain=dtrain,
   num_boost_round=n,
   evals=evals,
   verbose_eval=50,
   # Activate early stopping
   early_stopping_rounds=50
)

[0]	train-rmse:2859.49097	validation-rmse:2851.62630


/usr/local/lib/python3.10/dist-packages/xgboost/core.py:160: UserWarning: [03:57:45] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.

    E.g. tree_method = "hist", device = "cuda"

  warnings.warn(smsg, UserWarning)


[50]	train-rmse:430.36010	validation-rmse:532.90320
[100]	train-rmse:377.56825	validation-rmse:532.79980
[103]	train-rmse:375.44970	validation-rmse:532.50220


In [13]:
# Cross-validation

params = {"objective": "reg:squarederror", "tree_method": "gpu_hist"}
n = 1000

results = xgb.cv(
   params, dtrain,
   num_boost_round=n,
   nfold=5,
   early_stopping_rounds=20
)


/usr/local/lib/python3.10/dist-packages/xgboost/core.py:160: UserWarning: [03:57:47] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.

    E.g. tree_method = "hist", device = "cuda"

  warnings.warn(smsg, UserWarning)


In [14]:
results.head()

train-rmse-mean  train-rmse-std  test-rmse-mean  test-rmse-std
0      2861.153015        8.266765     2861.773555      36.937516
1      2081.378004        5.534608     2084.973481      32.064109
2      1545.361682        3.287745     1553.681211      31.059209
3      1182.364236        3.585787     1192.464771      26.157805
4       941.828819        2.971779      958.467497      23.613538

In [15]:
best_rmse = results['test-rmse-mean'].min()

best_rmse

549.1039652582465

## Start W&B


- Login into your W&B profile using the code below
- Alternatively you can set environment variables. There are several env variables which you can set to change the behavior of W&B logging. The most important are:
    - WANDB_API_KEY - find this in your "Settings" section under your profile
    - WANDB_BASE_URL - this is the url of the W&B server

- Find your API Token in "Profile" -> "Setttings" in the W&B App



In [16]:
# Log in to your W&B account
import wandb

wandb.login()

<IPython.core.display.Javascript object>

wandb: Logging into wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: You can find your API key in your browser here: https://wandb.ai/authorize
wandb: Paste an API key from your profile and hit enter, or press ctrl+c to quit:

 ··········


wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc


True

In [28]:
sweep_config = {
    'method': 'random',
    'metric': {
        'name': 'rmse',
        'goal': 'minimize'
    },
    'parameters': {
        'learning_rate': {
            'min': 0.01,
            'max': 0.2
        },
        'max_depth': {
            'values': [3, 5, 7, 9]
        },
        'colsample_bytree': {
            'min': 0.6,
            'max': 0.9
        },
        'n_estimators': {
            'values': [100, 150, 200, 250]
        }
    }
}


In [30]:
import wandb

sweep_id = wandb.sweep(sweep_config, project="diamonds_experiments", entity='shreyasi_jcx3419')

Create sweep with ID: 85bepaho
Sweep URL: https://wandb.ai/shreyasi_jcx3419/diamonds_experiments/sweeps/85bepaho


In [31]:
def train():
    run = wandb.init()

    config = wandb.config

    params = {
        'objective': 'reg:squarederror',
        'learning_rate': config.learning_rate,
        'max_depth': int(config.max_depth),
        'subsample': 0.9,
        'colsample_bytree': config.colsample_bytree,
        'n_estimators': int(config.n_estimators),
        'eval_metric': 'rmse'
    }

    model = xgb.train(params, dtrain, num_boost_round=config.n_estimators)

    predictions = model.predict(dtest)
    rmse = np.sqrt(mean_squared_error(y_test, predictions))

    wandb.log({'rmse': rmse})

    run.finish()


In [32]:
wandb.agent(sweep_id, train)

wandb: Agent Starting Run: kji4505m with config:
wandb: 	colsample_bytree: 0.6309207421453186
wandb: 	learning_rate: 0.03610392227790473
wandb: 	max_depth: 7
wandb: 	n_estimators: 100
wandb: Currently logged in as: shreyasi_jcx3419. Use `wandb login --relogin` to force relogin


wandb: Agent Starting Run: y4phi63w with config:
wandb: 	colsample_bytree: 0.8043782320341248
wandb: 	learning_rate: 0.02657517082673636
wandb: 	max_depth: 3
wandb: 	n_estimators: 100


wandb: Agent Starting Run: 7qedavsw with config:
wandb: 	colsample_bytree: 0.7195927256929011
wandb: 	learning_rate: 0.0375031642243228
wandb: 	max_depth: 7
wandb: 	n_estimators: 200


wandb: Agent Starting Run: ycrsirti with config:
wandb: 	colsample_bytree: 0.6550777607515582
wandb: 	learning_rate: 0.07937288235849038
wandb: 	max_depth: 3
wandb: 	n_estimators: 200


wandb: Agent Starting Run: i80io65p with config:
wandb: 	colsample_bytree: 0.644953499231781
wandb: 	learning_rate: 0.03114292955812417
wandb: 	max_depth: 5
wandb: 	n_estimators: 200


wandb: Agent Starting Run: sl47xymd with config:
wandb: 	colsample_bytree: 0.8721812952306391
wandb: 	learning_rate: 0.1016978301968217
wandb: 	max_depth: 5
wandb: 	n_estimators: 100


wandb: Agent Starting Run: ip6go59v with config:
wandb: 	colsample_bytree: 0.6218020275618733
wandb: 	learning_rate: 0.04815391261460225
wandb: 	max_depth: 7
wandb: 	n_estimators: 250


wandb: Agent Starting Run: 8e1bz8xr with config:
wandb: 	colsample_bytree: 0.7982189086824258
wandb: 	learning_rate: 0.14895782029958657
wandb: 	max_depth: 7
wandb: 	n_estimators: 200


wandb: Agent Starting Run: 15joznoo with config:
wandb: 	colsample_bytree: 0.86376521970067
wandb: 	learning_rate: 0.10245846789068146
wandb: 	max_depth: 9
wandb: 	n_estimators: 200


wandb: Agent Starting Run: fzzjbkju with config:
wandb: 	colsample_bytree: 0.6931333989259658
wandb: 	learning_rate: 0.0970773938791552
wandb: 	max_depth: 5
wandb: 	n_estimators: 150


wandb: Agent Starting Run: wdcu3b2w with config:
wandb: 	colsample_bytree: 0.8932619636557046
wandb: 	learning_rate: 0.14410028553741674
wandb: 	max_depth: 9
wandb: 	n_estimators: 250


Each entry in the log refers to different runs of the experiment, each with its own set of hyperparameters like `colsample_bytree`, `learning_rate`, `max_depth`, and `n_estimators`. Here's a brief overview of the results from these runs:

1. **Range of RMSE**: The RMSE values range from about 526.96 to 958.36, indicating varied performance across different configurations. Lower RMSE values signify better model performance, with the lowest observed RMSE being 526.96.
2. **Configuration Variations**: Each run tested different values for the hyperparameters. For example, `max_depth` varied from 3 to 9, `n_estimators` from 100 to 250, and learning rates and `colsample_bytree` also saw significant variations.
3. **Best Performing Runs**: The best performance was observed in the run with a `max_depth` of 7, `n_estimators` of 200, and `colsample_bytree` around 0.72. The worst-performing setup had a `max_depth` of 3, a `learning_rate` of around 0.027, and `n_estimators` set to 100.
4. **Link to Further Details**: Each run is linked to a detailed page on wandb, allowing further exploration of the configurations, performance graphs, and other metrics.
5. **Progress and Consistency**: The logs suggest ongoing testing and parameter tuning in an effort to optimize model predictions. Each run's data is saved, synchronized, and linked for easy access and analysis.
